Casting message to bytes in PyCrypto verifier.

dhermes opened the pull request:
Also
- Removing bare `except:` statements in both the PyCrypto and
  OpenSSL verifiers (in the `verify` method).
- Catching the only exception possible in the OpenSSL verifier
  (it is `OpenSSL.crypto.Error`).
- Converting the signature to bytes (if not already) in the
  OpenSSL verifier.
- Adding a test with both a unicode and bytes signature for
  each verifier.

Fixes #201.

/cc @nathanielmanistaatgoogle

dhermes commented:
@nathanielmanistaatgoogle @craigcitro As mentioned above, I removed the bare `except:` statements the verifiers. It's unclear if they were there intentionally or just done since the exception types were neglected.

nathanielmanistaatgoogle commented on tests/test_jwt.py:
```diff
@@ -59,7 +59,7 @@ def test_sign_and_verify(self):
     self._check_sign_and_verify('privatekey.%s' % self.format)
 
   def test_sign_and_verify_from_converted_pkcs12(self):
-    """Tests that following instructions to convert from PKCS12 to PEM works."""
```
What's up with the change on this line?

dhermes replied:
Docstrings on tests break the output of `nosetest` (and are generally frowned upon).

I can leave it out of this commit, I just noticed it because it showed up in a Py3.4 failure I fixed before adding

``` python
      if isinstance(signature, six.text_type):
        signature = signature.encode('utf-8')
```

to `OpenSSLVerifier.verify`.

nathanielmanistaatgoogle replied:
Wait, I've without knowing been doing something wrong for years _again_? Can you link me to a source for this general frowning?

dhermes replied:
Couldn't find anything in a few minutes of searching. I'm pretty sure I've seen something before.

Closest I got was the `unittest` docs for making the description ignore the docstring:
https://docs.python.org/2/library/unittest.html#unittest.TestCase.shortDescription

Maybe @tseaver has a quick reference

tseaver replied:
Docstrings in tnittest-based testcase methods are a bad idea because (historically, anyway) testrunners tend to display the docstring, rather than the name of the testcase, in output, making it hard to diagnose failures.

nathanielmanistaatgoogle commented on tests/test_jwt.py:
```diff
@@ -196,6 +198,7 @@ def setUp(self):
 
 
 class PEMCryptTestsOpenSSL(CryptTests):
+
```
Any comment on these blank line insertions?

dhermes replied:
Just consistency with `CryptTests`. I thought the extra line was a requirement for the linter, but after some experimenting it doesn't seem to be.

nathanielmanistaatgoogle commented on oauth2client/crypt.py:
```diff
@@ -93,12 +93,14 @@ def verify(self, message, signature):
         key that this object was constructed with.
       """
       from OpenSSL import crypto
+      if isinstance(message, six.text_type):
```
Any need to update the doc string for these arguments from "string"?

dhermes replied:
Yeah happy to do it.

craigcitro commented:
i don't know that there's a reason for the bare `except:` clauses -- but if there is, we'll probably get a bug report soon enough. :wink:

dhermes committed:
Casting message to bytes in PyCrypto verifier.

Also
- Removing bare `except:` statements in both the PyCrypto and
  OpenSSL verifiers (in the `verify` method).
- Catching the only exception possible in the OpenSSL verifier
  (it is `OpenSSL.crypto.Error`).
- Converting the signature to bytes (if not already) in the
  OpenSSL verifier.
- Adding a test with both a unicode and bytes signature for
  each verifier.

Fixes #201.

dhermes commented:
@nathanielmanistaatgoogle I have updated the docstrings on the updated `verify()` methods.

Sorry I let this thing fall away.

Are we good to go here?

nathanielmanistaatgoogle commented:
LGTM; I think we're good to go here.

dhermes commented:
Thanks!
